File: cpp/base/src/main/vsdk/toolkit/environment/scene/SimpleSceneSnapshot.cpp
#include "vsdk/toolkit/environment/scene/SimpleSceneSnapshot.h"
#include "vsdk/toolkit/environment/camera/CameraSnapshot.h"
#include "java/util/ArrayList.txx"

SimpleSceneSnapshot::SimpleSceneSnapshot(
    java::ArrayList<SimpleBody*>& inSimpleBodies,
    java::ArrayList<Light*>& inLights,
    Background* inBackground,
    CameraSnapshot* inCameraSnapshot)
    : background(inBackground), cameraSnapshot(inCameraSnapshot)
{
    for (long int i = 0; i < inSimpleBodies.size(); i++) simpleBodies.add(inSimpleBodies.get(i));
    for (long int i = 0; i < inLights.size(); i++) lights.add(inLights.get(i));
}

SimpleSceneSnapshot::~SimpleSceneSnapshot()
{
    if ( cameraSnapshot != 0 ) {
        delete cameraSnapshot;
        cameraSnapshot = 0;
    }
}

java::ArrayList<SimpleBody*>& SimpleSceneSnapshot::getSimpleBodies() { return simpleBodies; }
java::ArrayList<Light*>& SimpleSceneSnapshot::getLights() { return lights; }
Background* SimpleSceneSnapshot::getBackground() const { return background; }
CameraSnapshot* SimpleSceneSnapshot::getCameraSnapshot() const { return cameraSnapshot; }

File: typescript/base/src/main/vsdk/toolkit/environment/scene/SimpleSceneSnapshot.ts
import { ArrayList } from "../../../../java/util/ArrayList.js";
import { Collections } from "../../../../java/util/Collections.js";
import type { List } from "../../../../java/util/List.js";
import { Objects } from "../../../../java/util/Objects.js";

import { Entity } from "../../common/Entity.js";
import type { Background } from "../background/Background.js";
import type { CameraSnapshot } from "../camera/CameraSnapshot.js";
import type { Light } from "../light/Light.js";
import type { SimpleBody } from "./SimpleBody.js";

/**
Immutable scene container used by the raytracer for a consistent render pass.
*/
export class SimpleSceneSnapshot extends Entity {
    private readonly simpleBodies: List<SimpleBody>;
    private readonly lights: List<Light>;
    private readonly background: Background;
    private readonly cameraSnapshot: CameraSnapshot;

    public constructor(
        simpleBodies: List<SimpleBody>,
        lights: List<Light>,
        background: Background,
        cameraSnapshot: CameraSnapshot,
    ) {
        super();
        this.simpleBodies = Collections.unmodifiableList(
            new ArrayList<SimpleBody>(Objects.requireNonNull(simpleBodies, "simpleBodies cannot be null")),
        );
        this.lights = Collections.unmodifiableList(
            new ArrayList<Light>(Objects.requireNonNull(lights, "lights cannot be null")),
        );
        this.background = Objects.requireNonNull(background, "background cannot be null");
        this.cameraSnapshot = Objects.requireNonNull(cameraSnapshot, "cameraSnapshot cannot be null");
    }

    public getSimpleBodies(): List<SimpleBody> {
        return this.simpleBodies;
    }

    public getLights(): List<Light> {
        return this.lights;
    }

    public getBackground(): Background {
        return this.background;
    }

    public getCameraSnapshot(): CameraSnapshot {
        return this.cameraSnapshot;
    }
}
